Adds support for in-server service name resolution.

frankfarzan opened the pull request:
This is a continuation of PR #54163 authored by @filmil.

Note: the full design proposal is available at:

    https://goo.gl/APfsCE

This change allows webhook authorizer that call out to an endpoint of
the form:

    https://servicename.namespace.svc/someurl

to resolve "servicename.namespace.svc" to the service's current
endpoint, without contacting kube-dns.

This allows deployments without kube-dns access (e.g. GKE, or the
default k8s-on-GCP) to host the webhook authorizer as a pod.  It is
somewhat similar to the resolution already used for say admission
control plugins, but has a few important differences:

- The configuration files for admission control and webhook authz are
different.

- There is no dynamic registration support for authz webhooks.

The feature depends on a recent change that requires server certificate
to have CN subject field in the certificate be set to the Kubernetes
service name (for example: "/CN=servicename.namespace.svc").

Also, since the authz webhook is already released and being used, we
must uphold a few constraints:

- Deployments that don't need or use this feature should not have to pay
the cost of having it around.

- Since kube-apiserver is the only server that would need this feature,
the changes are confined to kube-apiserver specific code, and not pushed
deeper into generic webhook.

Changes inside:

- Added a custom dialer function that includes service resolution for
network addresses that can be interpreted as referring to a Kubernets
service.

- Some refactoring and reformatting of the existing code.  Namely the
webhook client construction was broken up into two smaller pieces:
reading the kubeconfig file, and creating the client proper from the
kubeconfig file.

- Handled code review comments.  This includes a rewrite of the original
ServiceResolver approach into a Dialer instead.  Per advice from
liggitt@.

<!--  Thanks for sending a pull request!  Here are some tips for you:
1. If this is your first time, read our contributor guidelines https://git.k8s.io/community/contributors/devel/pull-requests.md#the-pr-submit-process and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/devel/pull-requests.md#best-practices-for-faster-reviews
3. Follow the instructions for writing a release note: https://git.k8s.io/community/contributors/devel/pull-requests.md#write-release-notes-if-needed
-->

**What this PR does / why we need it**:
This pull request handles item (2) from #52511 It mirrors the approach that is already in use
for webhook admission control for example, but is adjusted to fit neatly in the current kubeconfig-based
webhook authz configuration.

**Which issue this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close that issue when PR gets merged)*: fixes #

**Special notes for your reviewer**:

**Release note**:
<!--  Steps to write your release note:
1. Use the release-note-* labels to set the release note state (if you have access)
2. Enter your extended release note in the below block; leaving it blank means using the PR title as the release note. If no release note is required, just write `NONE`.
-->
```release-note
Adds support for in-server service name resolution.

This change allows webhook authorizer that call out to an endpoint of
the form:

    https://servicename.namespace.svc/someurl

to resolve "servicename.namespace.svc" to the service's current
endpoint, without contacting kube-dns.

Design proposal is available at: https://goo.gl/APfsCE
```

frankfarzan committed:
Adds support for in-server service name resolution.

This is a continuation of PR #54163 authored by @filmil.

Note: the full design proposal is available at:

    https://goo.gl/APfsCE

This change allows webhook authorizer that call out to an endpoint of
the form:

    https://servicename.namespace.svc/someurl

to resolve "servicename.namespace.svc" to the service's current
endpoint, without contacting kube-dns.

This allows deployments without kube-dns access (e.g. GKE, or the
default k8s-on-GCP) to host the webhook authorizer as a pod.  It is
somewhat similar to the resolution already used for say admission
control plugins, but has a few important differences:

- The configuration files for admission control and webhook authz are
different.

- There is no dynamic registration support for authz webhooks.

The feature depends on a recent change that requires server certificate
to have CN subject field in the certificate be set to the Kubernetes
service name (for example: "/CN=servicename.namespace.svc").

Also, since the authz webhook is already released and being used, we
must uphold a few constraints:

- Deployments that don't need or use this feature should not have to pay
the cost of having it around.

- Since kube-apiserver is the only server that would need this feature,
the changes are confined to kube-apiserver specific code, and not pushed
deeper into generic webhook.

Changes inside:

- Added a custom dialer function that includes service resolution for
network addresses that can be interpreted as referring to a Kubernets
service.

- Some refactoring and reformatting of the existing code.  Namely the
webhook client construction was broken up into two smaller pieces:
reading the kubeconfig file, and creating the client proper from the
kubeconfig file.

- Handled code review comments.  This includes a rewrite of the original
ServiceResolver approach into a Dialer instead.  Per advice from
liggitt@.

ericchiang commented:
/ok-to-test

sttts commented on cmd/kube-apiserver/app/server.go:
```diff
@@ -445,7 +445,9 @@ func BuildGenericConfig(s *options.ServerRunOptions, proxyTransport *http.Transp
 	}
 
-	genericConfig.Authorizer, genericConfig.RuleResolver, err = BuildAuthorizer(s, sharedInformers)
+	genericConfig.Authorizer, genericConfig.RuleResolver, err =
```
ugly style. Better define local `dialer := buildWebhookDialer(serviceResolver, proxyTransport)` and keep it on one line.

sttts commented on cmd/kube-apiserver/app/server.go:
```diff
@@ -547,9 +549,55 @@ func BuildAuthenticator(s *options.ServerRunOptions, storageFactory serverstorag
+// buildWebhookDialer constructs a custom dialer function for the authorization webhook
+// that knows how to resolve references to Kubernetes services (of the form "servicename.namespace.svc")
+// to the IP address of the endpoint that currently implements it.
+func buildWebhookDialer(
```
newWebhookDialer

sttts commented on cmd/kube-apiserver/app/server.go:
```diff
@@ -547,9 +549,55 @@ func BuildAuthenticator(s *options.ServerRunOptions, storageFactory serverstorag
+		u, err := serviceResolver.ResolveEndpoint(namespace, service)
+		if err != nil {
+			return nil, fmt.Errorf(
+				"while resolving endpoint: namespace=%v, service=%v: %v",
```
don't we use `%s/%s` for namespace/object usually?

sttts commented on cmd/kube-apiserver/app/server.go:
```diff
@@ -547,9 +549,55 @@ func BuildAuthenticator(s *options.ServerRunOptions, storageFactory serverstorag
+	return func(network, address string) (net.Conn, error) {
+		host, _, err := net.SplitHostPort(address)
+		if err != nil {
+			glog.V(6).Infof("while splitting hostport: %v", err)
```
When can this fail? Do we understand less here than the baseDialer does?

liggitt replied:
no, I recommended delegating to the base dialer in this case

sttts commented on cmd/kube-apiserver/app/server.go:
```diff
@@ -547,9 +549,55 @@ func BuildAuthenticator(s *options.ServerRunOptions, storageFactory serverstorag
+		}
+		segments := strings.Split(host, ".")
+		if len(segments) != 3 {
+			glog.V(6).Infof("Dialer address is not a Kubernetes svc reference: %v", host)
```
Do we need this log output? Doesn't look like an error.

liggitt replied:
ends in .svc but is not a valid kubernetes reference... it's unexpected, I guess

sttts commented on pkg/kubeapiserver/options/authorization.go:
```diff
@@ -68,6 +71,13 @@ func (s *BuiltInAuthorizationOptions) AddFlags(fs *pflag.FlagSet) {
+	fs.BoolVar(&s.WebhookUseServiceResolution,
+		"authorization-webhook-use-service-resolution", s.WebhookUseServiceResolution,
+		"If set, the authorizer webhook code will resolve URLs pointing to hosts of the form "+
+			"https://servicename.namespace.svc:port/somewhere to appropriate endpoint IP addresses. "+
```
keep the indention as in the other flags please

ericchiang replied:
Other flags use `, ""+` to force the formatting.

https://play.golang.org/p/nU_aZ7Wd9M

frankfarzan replied:
Done

liggitt commented on cmd/kube-apiserver/app/server.go:
```diff
@@ -547,9 +549,55 @@ func BuildAuthenticator(s *options.ServerRunOptions, storageFactory serverstorag
+	}
+
+	return func(network, address string) (net.Conn, error) {
+		host, _, err := net.SplitHostPort(address)
```
realized this is dropping port info... we need to pass this to the service resolver, since different ports on a service can resolve to different endpoints

liggitt commented on pkg/kubeapiserver/authorizer/webhook.go:
```diff
@@ -0,0 +1,101 @@
+func (t *subjectAccessReviewClient) Create(
+	subjectAccessReview *authorization.SubjectAccessReview,
+) (*authorization.SubjectAccessReview, error) {
+	glog.V(2).Infof("Create request: %+v", subjectAccessReview)
```
agree, this is called extremely frequently... wouldn't really expect logging here

ericchiang commented on cmd/kube-apiserver/app/server_test.go:
```diff
@@ -0,0 +1,122 @@
+
+	for i, ttt := range tt {
+		var transport *http.Transport
+		defaultTransport := &http.Transport{}
```
This test still makes an arbitrary call to the internet. Please add a black hole dialer https://github.com/kubernetes/kubernetes/pull/54163#discussion_r147457488

frankfarzan committed:
Addressed reviewer comments:

- Fix Style issues.
- Disable spurious logging.
- Make tests hermetic by using dialer that always returns an error.

frankfarzan commented:
/retest

frankfarzan commented:
Unit test flaky issue: #54873

frankfarzan commented:
/retest

lavalamp commented:
I just learned from @bowei that this DNS name system for services is not actually allowed :( :(

Some discussion at https://github.com/kubernetes/kubernetes/pull/54889/files#r148385126

frankfarzan commented:
@lavalamp The design doc for this PR (https://goo.gl/APfsCE) discusses some of the options that were brought up in that thread.

I think the major difference here is that authz webhook uses a kubeconfig file for configuration so api-server needs to parse clusters[0].cluster.server to determine whether it's an in-cluster or external service. I think given the similarity between admission and authz webhook mechanism, there probably needs to be a larger effort to unify these surface areas, but this PR has a much more limited scope to fix resolving in-cluster services.

Also note that the service resolution in this PR is behind a flag, so in order to get ambiguity 1) Flag needs to be flipped 2) .svc tld needs to exist 3) a webhook needs to be served from .svc.

frankfarzan commented:
Discussion in PR #54889 is moving towards the direction of not using name.namespace.svc URL for dynamic admission controller webhook. These are the options I see here:

**option 1:** Use a custom scheme (e.g. k8s-service://servicename.namespace.svc/path) instead.
**option 2:** Add service as first class citizen to the authz webhook config file. This mirrors the approach taken by admission controller. However this requires either somehow adding new fields to kubeconfig or pigeonholing into existing fields: https://kubernetes.io/docs/admin/authorization/webhook/#configuration-file-format

What do you folks think? I'm not a K8S expert, so it's entirely possible that there are better options.

lavalamp commented:
I tentatively lean toward 2. @liggitt / @deads2k?

lavalamp commented:
Is the authz plugin expected to be used by extension apiservers? It might make a difference.

frankfarzan commented:
One way to formulate option 2:

```
# clusters refers to the remote service.
clusters:
  - name: name-of-remote-authz-service
    cluster:
      certificate-authority: /path/to/ca.pem      # CA for verifying the remote service.
      server: my-svc # Name of the webhook service

# users refers to the API Server's webhook configuration.
users:
  - name: name-of-api-server
    user:
      client-certificate: /path/to/cert.pem # cert for the webhook plugin to use
      client-key: /path/to/key.pem          # key matching the cert

# kubeconfig files require a context. Provide one for the API Server.
current-context: webhook
contexts:
- context:
    cluster: name-of-remote-authz-service
    user: name-of-api-server
    namespace: my-ns # Namespace of the webhook service
  name: webhook
```

If context.namespace is present, cluster.server represents the service name.

@liggitt @deads2k @sttts @ericchiang What do you think?

bgrant0607 commented:
Yes, please, option 2.

liggitt commented:
I don't necessarily object to making an explicit service-targeting option in kube-apiserver, but not by changing the meaning of the kubeconfig cluster and current context fields. If you want the kube-apiserver to be configured to contact a service that is inexpressible as a hostname, then you need a different config format.

Note that a more fit-to-purpose config format for the webhook is not a bad thing... the current one does not allow us to transition to v1 subjectaccessreviews, for example.

frankfarzan commented:
@liggitt: are you suggesting creating a yaml file config similar to ExternalAdmissionHookConfiguration? e.g.:

```
apiVersion: authorization.k8s.io/v1beta1
kind: ExternalAuthorizationHookConfiguration
metadata:
  name: example-config
externalAuthorizationHooks:
  clientConfig:
    serverCa: <ca for verifying the remote service>
    clientCertificate: <cert for the webhook plugin to use>
    clientKey: <key matching the cert>
    service:
      name: <name of the front-end service>
      namespace: <namespace of the front-end service>
    url: <URL of the remote webhook>
```

Is there a precedence for api-server parsing such files?

FWIW, kubeconfig file for authz has already changed the semantics of the fields such as cluster and user. I don't have context why kubeconfig was chosen as config format in the first place.

liggitt commented:
> @liggitt: are you suggesting creating a yaml file config similar to  ExternalAdmissionHookConfiguration?

Probably so. The further we get from a simple "connect to this URL with these credentials", the less the config matches a kubeconfig file, and the more a fit-to-purpose config struct is required

> Is there a precedence for api-server parsing such files?

Yes, the api server parses `--admission-control-config-file` (e.g. https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/apiserver/pkg/apis/apiserver/v1alpha1/types.go#L27)

> FWIW, kubeconfig file for authz has already changed the semantics of the fields such as cluster and user.

That's not really accurate. The standard kubeconfig loader is used to create a `restclient.Config` from the kubeconfig file, using the auth info from the user stanza, and server CA and URL from the cluster stanza:

https://github.com/kubernetes/kubernetes/blob/9ed5d380aa7fbe0f0461f4685d011030270a39e1/staging/src/k8s.io/apiserver/pkg/util/webhook/webhook.go#L57-L64

> I don't have context why kubeconfig was chosen as config format in the first place.

Because it was a standard structure within kube for describing a server URL, CA, and credentials for contacting a remote server, with pre-existing loaders and tranformers to construct transports/clients from it

The scheme proposed in https://github.com/kubernetes/kubernetes/pull/54733#issuecomment-342691129 changes the meaning of the `server` field to mean something completely different. If we want to be able to explicitly reference a service, not a URL, then we should have a config that matches that.

frankfarzan commented:
This PR is obsolete. Tentative plan is create a new config format as discussed above.
